File: checks/python/src/section508/checks/no_change_on_focus.py
import re

from section508.dom import attr, select

NAVIGATE = re.compile(r"location\s*=|window\.open|document\.location|href\s*=", re.I)


def check(root):
    if not root:
        return []

    for control in select(root, "select, input, textarea, button, a, [tabindex], [onfocus]"):
        handler = attr(control, "onfocus") or ""
        if NAVIGATE.search(handler):
            return ["Focus on this control would change context."]
    return []

File: checks/ruby/lib/section508/checks/no_change_on_focus.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Focusing a control must not navigate or rebuild the page.
    class NoChangeOnFocus
      NAVIGATE = /location\s*=|window\.open|document\.location|href\s*=/i

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        Dom.elements(root, "select, input, textarea, button, a, [tabindex], [onfocus]").each do |control|
          handler = control["onfocus"].to_s
          return ["Focus on this control would change context."] if NAVIGATE.match?(handler)
        end
        []
      end
    end
  end
end
